### PR TITLE
fix: detail-page back arrow should return to the originating list (#1299)

### LINKED ISSUE #1299: Detail-page back arrow should return to the originating list, not always the atlas page
## Summary

The back arrow on the **source dataset**, **integrated object**, and **source study** detail pages always lands on the atlas detail page (via the `BackButton` URL-segment fallback), regardless of where the user came from. It should return to the originating list — including the new global top-nav lists — and fall back to the previous breadcrumb step for direct/deep links.

## Repro

1. Open the global source datasets list: `/source-datasets`.
2. Click into a row → detail page at `/atlases/[atlasId]/source-datasets/[id]`.
3. Click the back arrow.
4. **Expected:** return to `/source-datasets`.
5. **Actual:** lands on the atlas detail page.

Same shape for:
- `/integrated-objects` → integrated object detail
- `/source-studies` → source study detail

…and symmetrically when entering from the atlas-scoped lists (`/atlases/[atlasId]/source-datasets`, `…/integrated-objects`, `…/source-studies`) — today the back arrow happens to land in roughly the right place only because `BackButton` drops the last URL segment, which is incidental rather than intentional.

## Root cause

Three detail views render `<DetailView />` without a `backPath`:

- `app/views/AtlasSourceDatasetView/atlasSourceDatasetView.tsx`
- `app/views/ComponentAtlasView/componentAtlasView.tsx`
- `app/views/SourceStudyView/sourceStudyView.tsx`

`BackButton` (`app/components/Layout/components/Detail/components/DetailViewHero/components/BackButton/backButton.tsx`) then falls back to a path-segment trim that is unaware of the actual origin.

## Proposed behavior

The back arrow should resolve in this order:

1. If the user arrived from a list (atlas-scoped or global), return to that list.
2. Otherwise (deep link / direct entry), return to the **previous breadcrumb step**, not the path-segment fallback.

## Prior art

PR #1206 (\`fix: validation summary back arrow returns to originating list\`) solved the analogous problem one level deeper, on the validation views, using:

- a hidden \`?from=list\` query threaded via Next.js \`Link href={{ pathname, query }}\`
- \`useRef\` to capture the origin on first render so it survives tab navigation
- a per-view \`backPath\` computed from that flag

This issue is the same mechanism extended one level up, with one new requirement: origin now distinguishes **atlas-scoped lists** from the new **global top-nav lists** added in #1269, #1270, #1271 — so a single boolean is no longer sufficient. Suggest \`from=atlas-list\` / \`from=global-list\` (or similar), and migrate the existing \`from=list\` usage to the new vocabulary in the same pass.

## Scope

Application code:
- The three detail views above — compute and pass \`backPath\`.
- A small shared helper (e.g. \`resolveBackPath(origin, pathParameter, breadcrumbs)\`).
- Every in-app link that targets these detail pages — add the appropriate \`from\` query. At minimum:
  - global lists: \`pages/source-datasets\`, \`pages/integrated-objects\`, \`pages/source-studies\`
  - atlas-scoped lists: \`pages/atlases/[atlasId]/{source-datasets,integrated-objects,source-studies}\`
  - any cross-links (e.g. from source-study source-datasets list, validation summaries)

Tests:
- View tests asserting the \`backPath\` matches the origin (atlas-list, global-list, deep-link).
- Update any existing test that relied on the URL-segment fallback.

## Edge cases

- **Tab switching:** capture origin once with \`useRef\` so it survives tab navigation (already the pattern from #1206).
- **Refresh on the detail page:** \`from\` query is lost → treat as deep link → previous breadcrumb step.
- **External referrers:** no \`from\` → deep-link fallback (correct).
- **Browser back:** unchanged; this issue is only about the in-app back arrow.

## Open questions

1. Origin vocabulary: \`from=atlas-list\` / \`from=global-list\` vs. keep \`from=list\` and disambiguate by URL inspection. Recommend the explicit form.
2. Centralize resolution in \`BackButton\` (read \`router.query.from\` there) vs. per-view computation. Recommend per-view for symmetry with the validation-view fix.
3. Should this also extend to other detail pages (user, metadata entry sheet validations)? Probably out of scope.

### PR DESCRIPTION
**Pull request by frano-m**
## Summary

Fixes the back arrow on source-dataset, integrated-object, and source-study detail pages (plus their validation views) so it returns to the originating list — global or atlas-scoped — instead of falling through to a URL-segment trim that often landed users in the wrong place.

## Mechanism

- Row links to detail pages now carry a `?from=<ROUTE_KEY>` query (the ROUTE key naming the originating list).
- New `useBackPath(pathParameter)` hook snapshots `query.from` via `useRef` at first render (so a tab switch within the detail view doesn't repoint the back arrow), then resolves the back path by looking up `ROUTE[origin]` and substituting the current `pathParameter`.
- Five detail views + three atlas-scoped list views adopt the hook and pass `backPath` into `<DetailView>` / `<StyledDetailView>`.
- `ValidationStatusCell` + `ValidationSummary` gain a `backOrigin: BackOrigin` prop so the same cell rendered in atlas-scoped vs global lists emits the right `from` query. This migrates away from the old hardcoded `from=list` introduced in #1206.

## Origin vocabulary

`BackOrigin = keyof typeof ROUTE` (aliased from the existing `RouteKey` type). Origins used:

- `ATLASES` — `/atlases` count cells (source-datasets / integrated-objects / source-studies columns)
- `SOURCE_DATASETS`, `SOURCE_STUDIES`, `INTEGRATED_OBJECTS` — top-nav global lists
- `ATLAS_SOURCE_DATASETS`, `ATLAS_SOURCE_STUDIES`, `COMPONENT_ATLASES` — atlas-scoped lists
- `METADATA_ENTRY_SHEETS` — source-study links from the metadata-entry-sheets table
- `REPORTS` — source-study links from the reports task table

## Resolution

`resolveBackPath` calls `getRouteURL(ROUTE[origin], pathParameter)` inside a try/catch — unsatisfiable origin (e.g. a sub-list route whose dynamic segments aren't in the destination detail page's `pathParameter`) falls through to `undefined`, letting `BackButton`'s existing URL-segment trim run. So deep links / direct URL entries behave exactly as before this PR.

## Sub-lists deliberately untouched

`IntegratedObjectSourceDatasetsView` and `ATLAS_SOURCE_STUDY_SOURCE_DATASETS` row links aren't wrapped — their natural origin route includes `[componentAtlasId]` / `[sourceStudyId]` not present in the source-dataset detail page's `pathParameter`. Behaviour matches `main` (URL-trim fallback). Filed as future work if needed.

## Tests

Two new test files, **19 tests, all pass**:

- `__tests__/back-button-utils.test.ts` — 14 pure-function assertions covering `parseBackOrigin`, `resolveBackPath` (including the unsatisfiable-segment catch case), and `withBackOrigin`.
- `__tests__/use-back-path.test.ts` — 5 hook tests with mocked `next/router`, including the snapshot-stability case (proves `useRef` lock survives a `query.from` change mid-lifecycle).

## Test plan

See [test plan in the conversation](https://github.com/clevercanary/hca-atlas-tracker/issues/1299) — 17 manual scenarios covering each origin × detail view, validation views, tab persistence, deep-link fallback, and edge cases (unknown `from`, browser back/forward).

Headline checks:
- [x] From `/source-datasets` → row → back arrow returns to `/source-datasets`
- [x] From `/atlases/X/source-datasets` → row → back arrow returns to `/atlases/X/source-datasets`
- [x] From `/atlases` → count cell → back arrow returns to `/atlases`
- [x] From `/reports` → source-study link → back arrow returns to `/reports`
- [x] From an atlas metadata-entry-sheets table → source-study link → back arrow returns to that table
- [x] Tab switch on a detail page doesn't change where the back arrow points
- [x] Deep-linked URL with no `from` → back arrow behaves as it did on `main`

Closes #1299

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: app/views/AtlasSourceDatasetValidationView/atlasSourceDatasetValidationView.tsx [hunk 1]
```diff
@@ -25,19 +23,17 @@ interface Props {
...
   const { sourceDataset } = useFetchAtlasSourceDataset(pathParameter);
   const formManager = useFormManager();
   const {
     access: { canView },
     isLoading,
   } = formManager;
-  const backPath = fromList.current
-    ? getRouteURL(ROUTE.ATLAS_SOURCE_DATASETS, pathParameter)
-    : getRouteURL(ROUTE.ATLAS_SOURCE_DATASET, pathParameter);
+  const backPath = useBackPath(pathParameter);
```
[the remark this answers is not in the document]
Comment (frano-m) [reply app/views/AtlasSourceDatasetValidationView/atlasSourceDatasetValidationView.tsx]: Good catch — restored the deep-link fallback to `ROUTE.ATLAS_SOURCE_DATASET` so a deep-linked validator page returns to the source-dataset detail (not URL-trim → validator-index which redirects to `cap`). Pushed in 864b878d.

File: app/views/IntegratedObjectValidationView/integratedObjectValidationView.tsx [hunk 2]
```diff
@@ -25,19 +23,17 @@ interface Props {
...
   const {
     access: { canView },
     isLoading,
   } = formManager;
-  const backPath = fromList.current
-    ? getRouteURL(ROUTE.COMPONENT_ATLASES, pathParameter)
-    : getRouteURL(ROUTE.COMPONENT_ATLAS, pathParameter);
+  const backPath = useBackPath(pathParameter);
+
   if (isLoading) return <Fragment />;
```
[the remark this answers is not in the document]
Comment (frano-m) [reply app/views/IntegratedObjectValidationView/integratedObjectValidationView.tsx]: Same fix applied here — deep-link fallback is now `ROUTE.COMPONENT_ATLAS`. Pushed in 864b878d.

File: app/components/Layout/components/Detail/components/DetailViewHero/components/BackButton/hooks/UseBackPath/utils.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+/**
+ * Parses an unknown `from` query value into a known back-origin (a `ROUTE`
+ * key), or `undefined` if the value isn't a recognised key.
+ * @param value - Raw `from` query value (from `useRouter().query.from`).
+ * @returns Recognised back origin or undefined.
+ */
+export function parseBackOrigin(value: unknown): BackOrigin | undefined {
+  if (typeof value === "string" && value in ROUTE) return value as BackOrigin;
+  return undefined;
```
[the remark this answers is not in the document]
Comment (frano-m) [reply app/components/Layout/components/Detail/components/DetailViewHero/components/BackButton/hooks/UseBackPath/utils.ts:48]: Fixed in 864b878d — `parseBackOrigin` now uses `Object.prototype.hasOwnProperty.call(ROUTE, value)` to reject inherited keys (`__proto__`, `toString`, etc.) coming from untrusted router query. Test coverage added in `back-button-utils.test.ts`.

File: app/components/Layout/components/Detail/components/DetailViewHero/components/BackButton/utils.ts [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+ * Appends a `from=<origin>` query parameter to a URL so the target view can
+ * resolve its back-arrow against the originating list. Centralises the query-
+ * string format used by all row links into the source-dataset, integrated-
+ * object, source-study, and validation detail pages.
+ * @param url - Target detail-page URL.
+ * @param origin - Back origin (a `ROUTE` key) to record.
+ * @returns URL with `?from=<origin>` appended.
+ */
+export function withBackOrigin(url: string, origin: BackOrigin): string {
+  return `${url}?from=${origin}`;
```
[the remark this answers is not in the document]
Comment (frano-m) [reply app/components/Layout/components/Detail/components/DetailViewHero/components/BackButton/utils.ts]: Fixed in 864b878d — `withBackOrigin` now goes through the URL API (`new URL(url, DUMMY_BASE)` + `searchParams.set`) so an existing query string is joined with `&` and any hash fragment is preserved. Dummy base is required because `URL` rejects relative paths; we return `pathname + search + hash` so the base never leaks. Test coverage added covering existing query, hash, and both combined.

File: __tests__/use-back-path.test.ts [hunk 5]
```diff
@@ -0,0 +1,71 @@
+import { renderHook } from "@testing-library/react";
+
+jest.mock("next/router", () => ({
+  useRouter: jest.fn(),
+}));
+
+// eslint-disable-next-line @typescript-eslint/no-require-imports -- jest.mock requires require()-style import after mock is set up.
```
[the remark this answers is not in the document]
Comment (frano-m) [reply __tests__/use-back-path.test.ts]: Good catch — you're right, that was a regular ES import not a `require()`. The disable was inert and the comment was misleading. Removed in 100cb417.
